clients/command/src/main/java/org/eclipse/hono/client/command/AbstractCommandContext.java

```diff
@@ -0,0 +1,179 @@
+/**
+ * Copyright (c) 2023 Contributors to the Eclipse Foundation
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * https://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.eclipse.hono.client.command;
+
+import java.net.HttpURLConnection;
+import java.util.Collections;
+import java.util.Objects;
+import java.util.Optional;
+
+import org.eclipse.hono.tracing.TracingHelper;
+import org.eclipse.hono.util.CommandConstants;
+import org.eclipse.hono.util.MapBasedExecutionContext;
+import org.eclipse.hono.util.MessagingType;
+import org.eclipse.hono.util.RegistrationAssertion;
+import org.eclipse.hono.util.TenantObject;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import io.opentracing.Span;
+import io.opentracing.tag.Tags;
+import io.vertx.core.Future;
+import io.vertx.core.json.JsonObject;
+
+/**
+ * A base class providing utility methods for passing around parameters relevant for processing a {@code Command} used
+ * in a Pub/Sub and Kafka based command context.
+ */
+public abstract class AbstractCommandContext extends MapBasedExecutionContext implements CommandContext {
```

Comment:
How about parameterizing this class with the concrete Command and CommandResponseSender types so that we do not need the type checks in the other  classes?

Reply:
Sure! Parameterized it with Command for now, since we only needed the type checks for the command. But if you think we should also parameterizing it with the CommandResponseSender I can update it as well